[QA-1571] adding additional logging for createResource, foreign key violation

andy7i opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/QA-1571
<Put notes here to help reviewer understand this PR>

Adds a bit more logging to help with debugging (in general, and for QA-1571)


foreign key violation error message example (from local):
```
{"causes":[],"message":"Foreign Key Violation(s) while deleting group(s): List(Map(groupId -> 7238732, groupName -> billing-project_aj-ca-1206_owner, still used in group(s): -> {workspace_1b86b9cd-35f2-4867-ad53-54be57a65856_project-owner,workspace_77fa1cce-b004-45c9-8429-da403cf905eb_project-owner,workspace_9845724b-34d9-40ca-9452-d391677ff277_project-owner}))","source":"sam","stackTrace":[],"statusCode":500}%
```

example of new createResource:
```
{"@timestamp":"2021-09-21T21:06:01.389Z","@version":"1","message":"Creating new `billing-project` with resourceId: `gpalloc-qa-master-xnh3dtz`","logger_name":"org.broadinstitute.dsde.workbench.sam.service.ResourceService","thread_name":"scala-execution-context-global-81","severity":"INFO","level_value":20000}

```

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

andy7i committed:
adding additional logging for createResource

andy7i committed:
adding additional logging for foreign key violation errors when deleting a resource's groups

andy7i committed:
add possibly necessary parentheses, remove unneeded type conversion

andy7i committed:
use for comps and Either

andy7i committed:
minor text fixes

andy7i commented on src/main/resources/reference.conf:
```diff
@@ -53,7 +53,7 @@ resourceTypes = {
       }
       "share_policy::can-compute" = {
-        description = "change the member of the can-compute policy for this workspace"
+        description = "change the membership of the can-compute policy for this workspace"
```
minor text fix

andy7i commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -744,23 +744,53 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
     val g = GroupTable.syntax("g")
+    val gm = GroupMemberTable.syntax("gm")
 
-    serializableWriteTransaction("deletePolicy", samRequestContext)({ implicit session =>
```
moved this write transaction into a for comp.
this write transaction was split into 1 read transaction and 1 write transaction. this is so that we can still use groupPKsToDelete if the write transaction fails

andy7i commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:
```diff
@@ -102,14 +102,16 @@ class ResourceService(
     */
-  def createResource(resourceType: ResourceType, resourceId: ResourceId, policiesMap: Map[AccessPolicyName, AccessPolicyMembership], authDomain: Set[WorkbenchGroupName], parentOpt: Option[FullyQualifiedResourceId], userId: WorkbenchUserId, samRequestContext: SamRequestContext): IO[Resource] =
+  def createResource(resourceType: ResourceType, resourceId: ResourceId, policiesMap: Map[AccessPolicyName, AccessPolicyMembership], authDomain: Set[WorkbenchGroupName], parentOpt: Option[FullyQualifiedResourceId], userId: WorkbenchUserId, samRequestContext: SamRequestContext): IO[Resource] = {
+    logger.info(s"Creating new `${resourceType}` with resourceId: `${resourceId}`")
```
tech debt item -- it's difficult to find the createResource call made to sam.

andy7i committed:
just need the name, not the entire encyclopedia

```
{"@timestamp":"2021-09-21T21:06:01.389Z","@version":"1","message":"Creating new `ResourceType(billing-project,HashSet(ResourceActionPattern(alter_policies,create and delete policies for this billing-project,false), ResourceActionPattern(list_children,list children, e.g. google-project,false), ResourceActionPattern(share_policy\"::.+\",modify the membership of the specified policy,false), ResourceActionPattern(read_spend_report_configuration,read the spend report config (i.e. the BigQuery dataset for spend reports) for this billing-project,false), ResourceActionPattern(list_notebook_cluster,list all notebook clusters in this billing-project,false), ResourceActionPattern(delete,delete this billing-project,false), ResourceActionPattern(view_status,view status of a billing-project,false), ResourceActionPattern(list_persistent_disk,list all persistent disks in this billing-project,false), ResourceActionPattern(remove_child,remove children, e.g. google-project,false), ResourceActionPattern(create_kubernetes_app,create kubernetes application in this billing-project,false), ResourceActionPattern(delete_notebook_cluster,,false), ResourceActionPattern(update_billing_account,update the billing account on this billing-project,false), ResourceActionPattern(delete_persistent_disk,delete persistent disk in this billing-project,false), ResourceActionPattern(stop_start_notebook_cluster,stop and start notebook clusters in this billing-project,false), ResourceActionPattern(launch_batch_compute,launch workflows in this billing-project,false), ResourceActionPattern(read_policies,list all policies and policy details for this billing-project,false), ResourceActionPattern(read_policy\"::.+\",view the membership of the specified policy,false), ResourceActionPattern(add_to_service_perimeter,add this billing-project to a service-perimeter,false), ResourceActionPattern(alter_spend_report_configuration,alter the spend report config (i.e. the BigQuery dataset for spend reports) for this billing-project,false), ResourceActionPattern(create_workspace,create new workspaces in this billing-project,false), ResourceActionPattern(alter_google_role,modify the role of users on the google project (only a fixed set of roles are permitted),false), ResourceActionPattern(create_persistent_disk,create persistent disk in this billing-project,false), ResourceActionPattern(launch_notebook_cluster,launch a new notebook cluster in this billing-project,false), ResourceActionPattern(add_child,add children, e.g. google-project,false)),Set(ResourceRole(owner,HashSet(list_children, alter_google_role, delete_notebook_cluster, read_policies, create_persistent_disk, create_workspace, delete, launch_batch_compute, read_spend_report_configuration, list_notebook_cluster, launch_notebook_cluster, create_kubernetes_app, list_persistent_disk, add_to_service_perimeter, view_status, remove_child, alter_spend_report_configuration, add_child, delete_persistent_disk, update_billing_account, alter_policies),Set(),Map(google-project -> Set(owner))), ResourceRole(notebook-user,Set(launch_notebook_cluster, create_persistent_disk, create_kubernetes_app),Set(),Map(google-project -> Set(notebook-user))), ResourceRole(batch-compute-user,Set(launch_batch_compute),Set(),Map()), ResourceRole(workspace-creator,Set(view_status, create_workspace, share_policy::can-compute-user, read_policy::can-compute-user),Set(),Map())),owner,true)` with resourceId: `gpalloc-qa-master-xnh3dtz`","logger_name":"org.broadinstitute.dsde.workbench.sam.service.ResourceService","thread_name":"scala-execution-context-global-81","severity":"INFO","level_value":20000}
```

gpolumbo-broad commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -744,23 +744,53 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+    for {
+      groupPKsToDelete <- readOnlyTransaction("deletePolicyFindGroups", samRequestContext) { implicit session =>
+          samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
+        }
```
I know the intention here was to split this into multiple transactions, but I'm not sure that's a good idea either.  Technically, if someone did something like adding a Policy to this resource between these transactions, then we would fail to properly clean up the Group behind that new Policy.  

If we had a cascading delete on the `sam_resource_policy.group_id` then I think we would be OK and we could just delete the policies and not have to worry about also deleting the groups, but that does not seem to be the case.  I do not know why the cascading delete is missing, so we would need to do some research before considering adding it.  

It seems like we should be able to read the `groupPKsToDelete` from within a single write transaction, but I don't know much about Slick and don't know what you were wrestling with, so I appreciate that it might not seem as straight forward as I think.

andy7i replied:
> Technically, if someone did something like adding a Policy to this resource between these transactions, then we would fail to properly clean up the Group behind that new Policy.

good point, that thought didn't occur to me.

> If we had a cascading delete on the sam_resource_policy.group_id then I think we would be OK and we could just delete the policies and not have to worry about also deleting the groups, but that does not seem to be the case. I do not know why the cascading delete is missing, so we would need to do some research before considering adding it.

you might be right about this. Based on the dev db, we have a 1 to 1 mapping of policies to distinct groups. 
```
sam=> select COUNT(*) from sam_resource_policy ; select COUNT(*) from (select distinct group_id from sam_resource_policy) as nq;
 602731

 602731
```
Maybe worth adding this to the tech debt backlog?


>It seems like we should be able to read the groupPKsToDelete from within a single write transaction, but I don't know much about Slick and don't know what you were wrestling with, so I appreciate that it might not seem as straight forward as I think.

If `groupPKsToDelete` is set inside the write transaction, and the write transaction succeeds, I think getting `groupPKsToDelete` is simple. However, we need `groupPKsToDelete` when the write transaction fails. Is there a way to get `groupPKsToDelete` if the write transaction fails?

marctalbott replied:
you could just get `groupPKstoDelete` again in the error handling transaction. Not the most efficient, but at this point things have already gone wrong so as long as it doesn't take down Sam it's not like users will mind if an error gets back to them slightly slower.

Agreed that it'd be nice to have the cascading delete, but like Greg I can't remember why we omitted it in the first place and I'd want to do a bit of research to be sure we aren't overlooking some important reason

gpolumbo-broad replied:
I know it's a pain, but I think it's worth keeping these queries in the same transaction, and following Marc's suggestion of rerunning the query when generating the error message.  Not ideal, but should work.

As for trying to get the intermediate result of `groupPKstoDelete`, I would try starting the transaction and then try using `zip` per https://scala-slick.org/doc/3.2.0/dbio.html#sequential-execution.  I'm not 100% that that will work for you, but that's where I'd start at least.

andy7i replied:
totally makes sense. I combined the transactions and pulled out the extracted the query into a val and now it is run twice.

marctalbott commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -744,23 +744,53 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+          for {
+            (problematicGroupIdsToNames, problematicGroupMemberships) <- readOnlyTransaction("deletePolicyErrorFindProblematicGroups", samRequestContext) { implicit session =>
+            val problematicGroupIdsToNames: Map[GroupPK, String] =
+              samsql"""select ${g.result.id}, ${g.result.name}
```
These queries feel like they could be condensed into a single query -- possibly something like this? While we're grabbing all the problematic group ids and names, also grab all of the groups that they're a member of.

```
select ${g.result.id}, ${g.result.name}, ${pg.result.name}
                     from ${GroupTable as g}
                     join ${GroupMemberTable as gm} on ${g.id} = ${gm.memberGroupId}
                     join ${GroupTable as pg} on ${pg.id} = ${gm.groupId}
                     where ${g.id} in
                         (select distinct ${gm.result.memberGroupId}
                          from ${GroupMemberTable as gm}
                          where ${gm.memberGroupId} in ($groupPKsToDelete))
```

gpolumbo-broad replied:
+1

gpolumbo-broad replied:
+1

marctalbott commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -744,23 +744,53 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
                  where ${g.id} in ($groupPKsToDelete)""".update().apply()
+        }
+      }).attempt
+      _ <- writeResult match {
```
It might be worth breaking this match out into a separate function to avoid the nested match and for comp

gpolumbo-broad commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -744,23 +744,53 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+        }
+      }).attempt
+      _ <- writeResult match {
+        case Left(fkViolation: PSQLException) if fkViolation.getSQLState == PSQLStateExtensions.FOREIGN_KEY_VIOLATION =>
```
In typical Greg fashion, I suggest extracting this case into its own method

gpolumbo-broad reviewed (commented):
I'm a little worried about the transaction change, but I'm not sure the problems you were seeing either, so maybe we need to talk about it some more.

andy7i committed:
pr feedback -- refactoring 2 sql queries to 1

andy7i committed:
pr feedback -- refactoring nested match and for comp

andy7i committed:
fix query

andy7i committed:
more abstraction

error message:
```
{"causes":[],"message":"Foreign Key Violation while deleting groups: List(Map(groupId -> 7238732, groupName -> billing-project_aj-ca-1206_owner, still used in group: -> workspace_1b86b9cd-35f2-4867-ad53-54be57a65856_project-owner), Map(groupId -> 7238732, groupName -> billing-project_aj-ca-1206_owner, still used in group: -> workspace_77fa1cce-b004-45c9-8429-da403cf905eb_project-owner), Map(groupId -> 7238732, groupName -> billing-project_aj-ca-1206_owner, still used in group: -> workspace_9845724b-34d9-40ca-9452-d391677ff277_project-owner))","source":"sam","stackTrace":[],"statusCode":500}%
```

andy7i committed:
cleanup

marctalbott commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -745,22 +745,64 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+        .map(rs =>
+          Map("groupId" -> rs.get[GroupPK](g.resultName.id).value.toString,
+            "groupName" -> rs.get[String](g.resultName.name),
+            "still used in group:" -> rs.get[String](pg.resultName.name)))
```
might be more readable if these were collected into a single list so the output looked something like 
`groupId -> 12, groupName -> 'example', still used in group: -> [34, 45, 56]` , but not blocking

andy7i replied:
updated:
```
{"causes":[],"message":"Foreign Key Violation(s) while deleting group(s): List(Map(groupId -> 7238732, groupName -> billing-project_aj-ca-1206_owner, still used in group(s): -> {workspace_1b86b9cd-35f2-4867-ad53-54be57a65856_project-owner,workspace_77fa1cce-b004-45c9-8429-da403cf905eb_project-owner,workspace_9845724b-34d9-40ca-9452-d391677ff277_project-owner}))","source":"sam","stackTrace":[],"statusCode":500}%
```

marctalbott commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -745,22 +745,64 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+    } yield ()
+  }
+
+  private def handleResult(samRequestContext: SamRequestContext, writeResult: Either[Throwable, AnyVal], groupPKsToDelete: List[GroupPK]) = {
```
could you add type annotations to these private functions?

marctalbott commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:
```diff
@@ -745,22 +745,64 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+  private def handleForeignKeyViolation(samRequestContext: SamRequestContext, groupPKsToDelete: List[GroupPK]) = {
+    for {
+      problematicGroups <- getGroupsCausingForeignKeyViolation(samRequestContext, groupPKsToDelete)
+      _ <- IO.raiseError[Unit](new WorkbenchExceptionWithErrorReport( // throws a 500 since that's the current behavior
```
Out of curiosity, why is `[Unit]` needed here but not when we're using `IO.raiseError` above?

andy7i replied:
I'm not sure why, but without the `[Unit]` the compiler says `dead code following this construct`.

```
[error] /app/src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala:782:9: dead code following this construct
[error]       _ <- IO.raiseError(new WorkbenchExceptionWithErrorReport( // throws a 500 since that's the current behavior
[error]         ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

I also get the same error if I put add a [Nothing]. (`IO.raiseError(e)` return type is IO[Nothing]). 

![Screen Shot 2021-09-23 at 9 29 38 AM](https://user-images.githubusercontent.com/5375000/134516194-3affdc74-ec2d-495e-a817-716e758eb2bc.png)

Seems like there's technically a type mismatch but in the rest of sam code it seems like we don't usually need to typecast it and it doesn't cause any issues.

andy7i committed:
PR feedback -- single transaction ; run groupPKsToDeleteQuery separately in error case

andy7i committed:
PR feedback -- aggregate groups for readability:

```
{"causes":[],"message":"Foreign Key Violation(s) while deleting group(s): List(Map(groupId -> 7238732, groupName -> billing-project_aj-ca-1206_owner, still used in group(s): -> {workspace_1b86b9cd-35f2-4867-ad53-54be57a65856_project-owner,workspace_77fa1cce-b004-45c9-8429-da403cf905eb_project-owner,workspace_9845724b-34d9-40ca-9452-d391677ff277_project-owner}))","source":"sam","stackTrace":[],"statusCode":500}%
```
